src/api/firebase.js

```diff
@@ -92,16 +93,60 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
 export async function updateItem(listId, itemId, checked) {
 	const itemRef = doc(db, listId, itemId);
 	const itemDoc = await getDoc(itemRef);
-	const { dateLastPurchased, totalPurchases } = itemDoc.data();
+	const { dateCreated, dateLastPurchased, dateNextPurchased, totalPurchases } =
+		itemDoc.data();
 
-	try {
-		return updateDoc(itemRef, {
-			dateLastPurchased: checked ? new Date() : dateLastPurchased,
-			totalPurchases: checked ? totalPurchases + 1 : totalPurchases,
-			checked: checked,
-		});
-	} catch (error) {
-		console.log('updateDoc error', error);
+	const today = new Date();
+
+	const checkedDateLastPurchased = checked ? today : dateLastPurchased;
+
+	if (checkedDateLastPurchased) {
+		const dateNextPurchasedAsDate = dateNextPurchased.toDate();
+
+		/**
+		 * Calculate the previous estimate for purchase
+		 * based on historical data
+		 */
+		const previousEstimate = getDaysBetweenDates(
+			checkedDateLastPurchased,
+			dateNextPurchasedAsDate,
+		);
+
+		/**
+		 * Calculate the days since the last transaction,
+		 * considering either date last purchased or the created date
+		 */
+		const daysSinceLastTransaction = getDaysBetweenDates(
+			today,
+			checkedDateLastPurchased || dateCreated,
```

Comment:
I think we have to be careful here and above. 

Here, first, the `dateCreated` fallback isn't needed because we're inside a `if(checkedDateLastPurchased)`, it's unreachable. 

More importantly, when this function runs when `checkedDateLastPurchased` gets set as `dateLastPurchased (Firebase.timestamp)`  instead of `today (Date)` (when `checked` is false) both `previousEstimate` and `daysSinceLastTransaction` will be undefined, because the Firestore.timestamp type doesn't have a `getTime` method. So `checkedDateLastPurchased` just needs to get set with `dateLastPurchased.toDate()`

Reply:
Good point @chase-cove . Is that `if` check for `if(checkDateLastPurchased)` necessary? @AslanTaheri @rdpfeifle is there a situation where it could be falsy?

Reply:
@mentalcaries The `if (checkDateLastPurchased)` condition is necessary, and we shouldn't remove it, at least not for now.

Additionally, we should keep this code as is `const checkedDateLastPurchased = checked ? today : dateLastPurchased`.

Unfortunately, the alternative mentioned by @chase-cove doesn't work. Using `dateLastPurchased.toDate()` causes the item **not** to be checked on the frontend, and `dateNextPurchased` is no longer associated with `dateLastPurchased`. Also, the `previousEstimate` and `daysSinceLastTransaction` are not returning "undefined" as **is**. They would only return "undefined" if we hadn't used the `const today = new Date();`

Regarding `dateCreated` we only used it as requested in the `calculateEstimate` function from the `@the-collab-lab/shopping-list-utils` module [here](https://github.com/the-collab-lab/shopping-list-utils/blob/main/src/calculateEstimate/index.ts), specifically for the `daysSinceLastTransaction` parameter. `DateCreated` is set above (we destructured it), and if you examine the `addItem` function, it's also a JS Date object.

For now, making changes to avoid reliance on `checked` would be a complex and time-consuming task involving modifications in multiple files, such as `List.jsx` and `ListItem.jsx`. We should consider creating a new issue for this, as it comes from previous issues and would require significant changes to the current logic.

Reply:
@mentalcaries as of the current logic of the app (from previous issues and ours), `dateLastPurchased` is only `null` (falsy) when we add a new item to the list for the first time. As soon as we click (i.e. "buy") an item, `the dateLastPurchased` is no longer falsy.

Reply:
Ok good. I'm just thinking of ways to avoid unreachable code as @chase-cove mentioned.
What if you flipped the logic and created an early return if `checkDateLastPurchased` was falsy: `if(!checkedDateLastPurchased)`